### ISSUE TITLE
Possible memory leak in addMolecules/subMolecules

### ISSUE THREAD

**Issue by stanstrup** [sneumann/Rdisop#1]
It looks like there is a memory leak in addMolecules/subMolecules. If you do:

``` R
elements=c(  initializePSE(),initializeCharges()   )

for (i in 1:8000){
     addMolecules("H38N32O15P8","H",elements=elements)$formula
}
```

Several hundred MB are consumed that are never released again (gc() doesn't help). Even if saving the output only takes 500 Kb.

**Comment by jalic** [comment]
Hi Stan, as Steffen already pointed out also get-functions are effected. My hypothesis is that in connection with C-code GarbageCollection doesn't work. Until it is fixed you can either substitute by your own function or use the equivalent functions from the enviPat package. But careful, enviPat requires explicit C1H4 instead of CH4, so you have to use enviPat::check_chemform before processing Rdisop output.

**Comment by janlisec** [comment]
@sneumann You are right, as `addMolecules` returns a molecules object which is internally generated by C++ code similar to what `getMolecule` uses the error occurs in both cases. I tried to track down the offending statements in C++ leading to the potential memory leak and noticed the following. The error occurs for the following minimal example (at my PC using 10^4 iterations, this might by more or less on other systems):

```
for (i in 1:10^4){
    Rdisop::addMolecules("C", "H")
}
```

I can avoid this error explicitely specifying the allowed elements:

```
for (i in 1:10^4){
    Rdisop::addMolecules("C", "H", elements = initializeElements(c("C","H")))
}
```

The latter runs for me till 10^6 without any problems.

After further checking I realized that it is probably the number of elements that causes the problem, the more elements are specified, the sooner R crashes. A possible workaround would be to limit the element list provided to C++ within the calling R functions (here: `addMolecules`) to the elements actually needed because they are present in the `formula` argument. What do you think?

**Comment by janlisec** [comment]
AddOn. After some further investigation, I would conclude that it is not the absolute number of elements but some specific ones with 7 isotopes in particular. I following example code runs without trouble for 10^5 iterations:

```
ele <- sapply(initializePSE(), function(x) { x$name })
n_isos <- setNames(sapply(initializePSE(), function(x) { length(x$isotope$mass) }), ele)
# use only elements with more or less isotopes than 7
elements <- initializeElements(names(which(n_isos!=7)))
for (i in 1:10^4) { Rdisop::getMolecule("C", elements = elements) }
```

while the counter check fails

```
# use only elements with exactly 7 isotopes (and add Carbon)
elements <- initializeElements(c("C", names(which(n_isos==7))))
for (i in 1:10^4) { Rdisop::getMolecule("C", elements = elements) }
```

Unfortunately, I can't pinpoint a single element with n_iso==7 to be the culprit. :(

**Comment by janlisec** [comment]
Issue solved in PR #35.

**Comment by janlisec** [comment]
@stanstrup Would you please close the issue or comment if it still does need work?

**Comment by stanstrup** [comment]
I cannot check the memory aspect of this as now my initial snippet crashes R after a few thousand runs.  v1.67.5 on windows 11.

**Comment by janlisec** [comment]
You are correct @stanstrup that your initial snippet crashes R. The reason for this I explained in 2 comments from 08.11.2024 above. As I am not sufficiently proficient in `C++` I can not identify why some elements with 7 isotopes cause the memory problem.

The fix of PR #35 allows to run your example code without initializing elements:
```
for (i in 1:8000) {
    Rdisop::addMolecules("H38N32O15P8","H")$formula
}
```
which works nicely, because elements are initiated based on the 2 provided formulas internally. Your original code explicitely includes the complete PSE. Therefore it crashes R.

@sneumann I would suggest to remove the `elements` parameter from `addMolecules()` to avoid users specifying elements causing a potential memory problem by accident (as in this issue). The only potential problem I see with this would be if people want to use `addMolecules()` in combination with user defined element lists -- but I would consider this rather academical. Let me know if I shall prepare a respective PR.

However, another issue may be opened some day by someone aiming to add a sum formula including one of the potentially offensive elements (Ce, Ge, Hf, Ni, W, Zn, Zr).

**Comment by stanstrup** [comment]
I can confirm that with your code there is no memory leak.

**Comment by sneumann** [comment]
Awesome! Thanks Jan and Jan ! Yours, Steffen
